controllers/providers/aws/ssm.go

```diff
@@ -0,0 +1,56 @@
+package aws
+
+import (
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ssm"
+	"github.com/aws/aws-sdk-go/service/ssm/ssmiface"
+	"github.com/keikoproj/aws-sdk-go-cache/cache"
+	"github.com/keikoproj/instance-manager/controllers/common"
+)
+
+const (
+	EksOptimisedAmiPath = "/aws/service/eks/optimized-ami/%s/amazon-linux-2/recommended/image_id"
+)
+
+func GetAwsSsmClient(region string, cacheCfg *cache.Config, maxRetries int, collector *common.MetricsCollector) ssmiface.SSMAPI {
+	config := aws.NewConfig().WithRegion(region).WithCredentialsChainVerboseErrors(true)
+	config = request.WithRetryer(config, NewRetryLogger(maxRetries, collector))
+	sess, err := session.NewSession(config)
+	if err != nil {
+		panic(err)
+	}
+	cache.AddCaching(sess, cacheCfg)
+	cacheCfg.SetCacheTTL("ssm", "GetParameter", GetParameterTTL)
+	sess.Handlers.Complete.PushFront(func(r *request.Request) {
+		ctx := r.HTTPRequest.Context()
+		log.V(1).Info("AWS API call",
+			"cacheHit", cache.IsCacheHit(ctx),
+			"service", r.ClientInfo.ServiceName,
+			"operation", r.Operation.Name,
+		)
+	})
+	return ssm.New(sess)
+}
+
+func (w *AwsWorker) GetEksLatestAmi(clusterName string) (string, error) {
+	log.Info("GetEksLatestAmi")
+	version, err := w.GetEksClusterVersion(clusterName)
```

Comment:
We should already have the cluster version since we make the call separately in CloudDiscovery.
Can you decouple it? e.g. take version as a parameter.

Reply:
I'll double check - I created this because I believe we call `DescribeCluster` for a different purpose

Reply:
My point is that we already make the DescribeCluster call on reconcile every time and it would be counter productive to make that call again (although we have caching) - we call it to get the cluster version actually.

So you can pass in the cluster from cached state into `GetEksLatestAmi` and avoid the extra API call.

https://github.com/keikoproj/instance-manager/blob/fe8b56f407f75ca71dbdb8e08892b6fece610e0a/controllers/provisioners/eks/cloud.go#L165-L169

You can use the following method on DiscoveredState:
https://github.com/keikoproj/instance-manager/blob/fe8b56f407f75ca71dbdb8e08892b6fece610e0a/controllers/provisioners/eks/cloud.go#L300-L305